io-sim/src/Control/Monad/IOSim/Internal.hs

```diff
@@ -925,17 +913,19 @@ execAtomically !time !tid !tlbl !nextVid0 action0 k0 =
           -- Return vars read, so the thread can block on them
           k0 $! StmTxBlocked $! (Map.elems read)
 
-        OrElseLeftFrame b k writtenOuter writtenOuterSeq createdOuterSeq ctl' ->
-          {-# SCC "execAtomically.go.OrElseLeftFrame" #-} do
-          -- Revert all the TVar writes within this orElse
+        BranchFrame (OrElseStmA b) k writtenOuter writtenOuterSeq createdOuterSeq ctl' ->
```

Comment:
I would suggest combining the two `BranchFrame` alteranatives, and then branching within that on the `OrElseStmA`. EG the `traverse ... written` could be de-duplicated that way.